scripts/xblock/xblock_counts.py

```diff
@@ -0,0 +1,386 @@
+    """
+    Totals the xBlock types included in the course data and returns those counts by type
+
+    Arguments:
+        course_data (list of dicts): a list of course_data objects
+
+    Returns:
+        dict: containing the total number of blocks by type
+            {
+                <block_type>: <count>,
+                ...
+            }
+        dict: containing the total unique courses for a block type
+    """
+    block_summary_counts = {}
+    unique_course_counts = {}
+
+    for course in course_data:
+        block_counts = course.get(BLOCK_COUNTS_KEY)
+        for count_label, value in block_counts.items():
+            unique = 0
+            if value > 0:
+                unique = 1
+            if count_label in block_summary_counts:
+                block_summary_counts[count_label] += value
+                unique_course_counts[count_label] += unique
+            else:
+                block_summary_counts[count_label] = value
+                unique_course_counts[count_label] = unique
+
+    return block_summary_counts, unique_course_counts
+
+
+def write_block_summary_report(course_data):
+    """
+    Generate a CSV file containing a summary of the xBlock usage
+
+    Arguments:
+        course_data (list of dicts): a list of course_data objects
+
+    Returns:
+        Nothing
+    """
+    (block_summary_counts, unique_course_counts) = _get_block_summary_totals(course_data)
+
+    # Open and start writing the data into the CSV
+    with open('xblock_summary_counts.csv', 'wb') as csvfile:
+        summary_writer = csv.writer(csvfile, delimiter=',',
+                                    quotechar='"', quoting=csv.QUOTE_MINIMAL)
+        summary_writer.writerow(['XBLOCK_NAME', 'UNIQUE_COURSES', 'NUM_TOTAL_INSTANCES'])
+        for block_type in sorted(block_summary_counts):
+            block_count = block_summary_counts.get(block_type)
+            summary_writer.writerow([block_type, str(unique_course_counts[block_type]), str(block_count)])
+        csvfile.close()
+
+
+def write_course_block_detail_report(course_data):
+    """
+    Generate a CSV file containing the detailed information about the xBlocks available per course
+
+    Arguments:
+        course_data (list of dicts): a list of course_data objects
+
+    Returns:
+        Nothing
+    """
+    with open('xblock_course_detail.csv', 'wb') as csvfile:
+        detail_writer = csv.writer(
+            csvfile,
+            delimiter=',',
+            quotechar='"',
+            quoting=csv.QUOTE_ALL
+        )
+        detail_writer.writerow(['XBLOCK_TYPE_NAME', 'COURSE_NAME', 'COURSE_ID', 'COURSE_START', 'COURSE_END', 'NUM_XBLOCK_INSTANCES'])
+        for course in course_data:
+            for block_type, count in course.get(BLOCK_COUNTS_KEY, []).items():
+                if count > 0:
+                    detail_writer.writerow([
+                        block_type,
+                        course.get(COURSE_NAME_KEY, '').encode('utf-8'),
```

Comment:
This is the only custom parameter. There is one course name with characters that cannot be encoded with ascii and the standard STR methods. If there is a better way to handle this I am open to changing it.